Add local storage (scratch space) allocatable support

jingxu97 opened the pull request:
This PR adds the support for allocatable local storage (scratch space).
This feature is only for root file system which is shared by kubernetes
componenets, users' containers and/or images. User could use
--kube-reserved flag to reserve the storage for kube system components.
If the allocatable storage for user's pods is used up, some pods will be
evicted to free the storage resource.

This feature is part of local storage capacity isolation and described in the proposal https://github.com/kubernetes/community/pull/306

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
This feature exposes local storage capacity for the primary partitions, and supports & enforces storage reservation in Node Allocatable 
```

jingxu97 committed:
Add local storage (scratch space) allocatable support

This PR adds the support for allocatable local storage (scratch space).
This feature is only for root file system which is shared by kubernetes
componenets, users' containers and/or images. User could use
--kube-reserved flag to reserve the storage for kube system components.
If the allocatable storage for user's pods is used up, some pods will be
evicted to free the storage resource.

dashpole reviewed (changes requested):
First pass.  I'll look into what changes to machineInfo it would take to be able to determine rootfs capacity.

jingxu97 commented on test/e2e_node/localStorageAllocatable_eviction_test.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// Pass podTestSpecsP as references so that it could be set up in the first BeforeEach clause
+func runLocalStorageEvictionTest(f *framework.Framework, conditionType v1.NodeConditionType, testCondition string, podTestSpecsP *[]podTestSpec, evictionTestTimeout time.Duration,
```
I tried a few things, only by passing the reference works for me. I plan to refactor this code in a separate PR, is that ok?

dashpole replied:
sure, that can happen after code freeze, if neccessary.

jingxu97 commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -83,6 +86,7 @@ func init() {
 	signalToResource[evictionapi.SignalNodeFsAvailable] = resourceNodeFs
 	signalToResource[evictionapi.SignalNodeFsInodesFree] = resourceNodeFsInodes
+	signalToResource[evictionapi.SignalAllocatableStorageScratchAvailable] = resourceNodeFs
 	resourceToSignal = map[v1.ResourceName]evictionapi.Signal{}
```
Yes, I will do it in a separate PR, basically changing from 1-1 mapping from resource to signal, we use map 1-many with a list of signals

jingxu97 commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -391,10 +391,13 @@ func (m *managerImpl) reclaimNodeLevelResources(resourceToReclaim v1.ResourceNam
 
 		// evaluate all current thresholds to see if with adjusted observations, we think we have met min reclaim goals
 		if len(thresholdsMet(m.thresholdsMet, observations, true)) == 0 {
+			glog.V(3).Infof("eviction manager: reclaim goal reached for resource %v: ", resourceToReclaim)
```
fixed

jingxu97 commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -197,20 +197,21 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
 		// this may error if cadvisor has yet to complete housekeeping, so we will just try again in next pass.
-		hasDedicatedImageFs, err := diskInfoProvider.HasDedicatedImageFs()
-		if err != nil {
+		if errImage != nil {
```
From the existing code, I think the logic is if resourceToRankFunc and resourceToNodeReclaimFuncs are already defined, it does not matter whether errImage returns nil or not. 
I modified the code here a little more. Instead of keeping calling HasDedicatedImageFs in every sync call, I put it in a map. PTAL

jingxu97 commented on pkg/kubelet/cm/node_container_manager.go:
```diff
@@ -180,9 +181,18 @@ func (cm *containerManagerImpl) getNodeAllocatableAbsolute() v1.ResourceList {
+// GetNodeAllocatable returns amount of compute or storage resource that have to be reserved on this node from scheduling.
 func (cm *containerManagerImpl) GetNodeAllocatableReservation() v1.ResourceList {
 	evictionReservation := hardEvictionReservation(cm.HardEvictionThresholds, cm.capacity)
+	if !hasStorageInfoInCapacity(cm.capacity) {
```
fixed

jingxu97 committed:
Add predicates check for local storage request

This PR adds the check for local storage request when admitting pods. If
the local storage request exceeds the available resource, pod will be
rejected.

jingxu97 commented on pkg/kubelet/kubelet_cadvisor.go:
```diff
@@ -100,3 +100,15 @@ func (kl *Kubelet) GetCachedMachineInfo() (*cadvisorapi.MachineInfo, error) {
 }
+
+// GetCachedRootFsInfo assumes that the rootfs info can't change without a reboot
+func (kl *Kubelet) GetCachedRootFsInfo() (cadvisorapiv2.FsInfo, error) {
```
thanks! I will check that too, but we might also need to add this after code freeze.

dashpole replied:
Yep, that is fine.

derekwaynecarr commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -504,13 +504,22 @@ func GetResourceRequest(pod *v1.Pod) *schedulercache.Resource {
+	//Account for storage requested by emptydir volumes
+	for _, vol := range pod.Spec.Volumes {
+		if vol.EmptyDir != nil {
+			result.StorageScratch += vol.EmptyDir.SizeLimit.Value()
```
this needs to exclude empty dir that is backed by memory.

jingxu97 replied:
fixed

jingxu97 replied:
fixed

derekwaynecarr commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -76,34 +76,55 @@ var (
+						Name: "emptyDirVolumeName",
+						VolumeSource: v1.VolumeSource{
+							EmptyDir: &v1.EmptyDirVolumeSource{
+								SizeLimit: *resource.NewQuantity(scratch, resource.BinarySI),
```
set medium to not memory

jingxu97 replied:
fixed

jingxu97 replied:
fixed

derekwaynecarr commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -384,7 +405,69 @@ func TestPodFitsResources(t *testing.T) {
+		reasons  []algorithm.PredicateFailureReason
+	}{
+		{
+			pod: newResourcePod(schedulercache.Resource{MilliCPU: 1, Memory: 1, StorageOverlay: 1}),
```
add a test for memory backed empty dir.

jingxu97 replied:
added a test for memory medium

derekwaynecarr commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -361,6 +366,14 @@ func calculateResource(pod *v1.Pod) (res Resource, non0_cpu int64, non0_mem int6
+	//Account for storage requested by emptydir volumes
+	for _, vol := range pod.Spec.Volumes {
+		if vol.EmptyDir != nil {
+			res.StorageScratch += vol.EmptyDir.SizeLimit.Value()
```
it would be good if we have a utility functoin for this as this loop is repeated in multiple places.

also need to ignore medium memory empty dir.

jingxu97 replied:
fixed. this loop only appeared in predicates.go and node_info.go, could we consider add a utility if more places need it?

vishh replied:
I'd imagine tests also using this utility function.

derekwaynecarr reviewed (changes requested):
requested a few changes.

jingxu97 commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -391,10 +391,13 @@ func (m *managerImpl) reclaimNodeLevelResources(resourceToReclaim v1.ResourceNam
 			continue
 		}
-		value.available.Add(*reclaimed)
+		if reclaimed != nil {
```
fixed

jingxu97 commented:
addressed the comments. @derekwaynecarr @davidopp PTAL

dashpole commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -211,21 +214,22 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
-		// this may error if cadvisor has yet to complete housekeeping, so we will just try again in next pass.
-		hasDedicatedImageFs, err := diskInfoProvider.HasDedicatedImageFs()
-		if err != nil {
+	if m.dedicatedImageFs == nil {
```
I think we still want to enter this if `len(m.resourceToRankFunc) == 0 || len(m.resourceToNodeReclaimFuncs) == 0`.
So this should be `if m.dedicatedImageFs == nil || len(m.resourceToRankFunc) == 0 || len(m.resourceToNodeReclaimFuncs) == 0`

dashpole replied:
JK, i am wrong.

dashpole commented:
eviction manager changes lgtm

vishh commented on test/e2e_node/localStorageAllocatable_eviction_test.go:
```diff
@@ -0,0 +1,322 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
This is an odd file name. I think go recommends using `_` instead of camelcase

jingxu97 replied:
I changed the file name

vishh commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -560,6 +560,17 @@ func (kl *Kubelet) setNodeStatusMachineInfo(node *v1.Node) {
+	if hasDedicatedImageFs, err := kl.HasDedicatedImageFs(); err != nil && hasDedicatedImageFs {
+		imagesfs, err := kl.ImagesFsInfo()
+		if err != nil {
+			node.Status.Capacity[v1.ResourceStorageOverlay] = resource.MustParse("0Gi")
```
Shouldn't we fail kubelet in this case? Why would kubelet lie that imagefs doesn't exist when it fails to collect stats about it?

jingxu97 replied:
I followed the logic in GetCachedMachineInfo. Should we use different behavior here?

vishh commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -76,9 +76,10 @@ type Resource struct {
+		v1.ResourceCPU:            *resource.NewMilliQuantity(r.MilliCPU, resource.DecimalSI),
+		v1.ResourceMemory:         *resource.NewQuantity(r.Memory, resource.BinarySI),
+		v1.ResourceNvidiaGPU:      *resource.NewQuantity(r.NvidiaGPU, resource.DecimalSI),
+		v1.ResourceStorageOverlay: *resource.NewQuantity(r.StorageOverlay, resource.BinarySI),
```
What about `StorageScratch`?

jingxu97 replied:
This is used in test which for resource request, we don't support StorageScratch as a resource request. Only StorageOverlay for container

vishh commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -349,6 +352,8 @@ func calculateResource(pod *v1.Pod) (res Resource, non0_cpu int64, non0_mem int6
 				res.Memory += rQuant.Value()
 			case v1.ResourceNvidiaGPU:
 				res.NvidiaGPU += rQuant.Value()
+			case v1.ResourceStorageOverlay:
```
What about `StorageScratch`?

vishh replied:
nvm I see it below

vishh reviewed (changes requested):
A couple of comments. Overall LGTM

vishh commented:
@jingxu97 Tie this PR to the parent feature issue and mark the parent feature issue as `approved for milestone`

derekwaynecarr commented:
assuming @vishh additional comments, this is good for me.

vishh commented:
/lgtm

vishh commented:
@jingxu97 squash your commits. The last commit name sounds dubious.

thockin commented:
/approve

falenn commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (185/150 (123%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 13223 characters long; only the first
------ 5000 or so are included here.
Received: from o10.sgmail.github.com ([167.89.101.201]:25989)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-5c7e-dev=imwiz.com@sgmail.github.com>)
	id 1dGTR4-00005N-VG
	for dev@imwiz.com; Thu, 01 Jun 2017 10:58:59 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=mZidIX3C8ud8vT8sLmJOXqPpVQU=; b=XSVLqG/qayZL+pr0
	4Ddox+gC5kMsqi4ltnBFxSI6cVOe/m470L1Ape669NDrZy6UsR7V2HW8gzegDk+5
	jA9tYFDNvR5gROF94aM6PuOtn4+AWGUzLv9w8EUog+F96yLnFWCnuNs4KziPcT1r
	R9jYA5d2EAVw07bpE2VrVbmhtsw=
Received: by filter0831p1mdw1.sendgrid.net with SMTP id filter0831p1mdw1-17748-593047B5-70
        2017-06-01 16:58:29.849551033 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0004p1iad1.sendgrid.net (SG) with ESMTP id JpfX9wQTSBmYZ48aZ2Kq4A
	for <dev@imwiz.com>; Thu, 01 Jun 2017 16:58:29.839 +0000 (UTC)
Date: Thu, 01 Jun 2017 09:58:29 -0700
From: k8s-ci-robot <notifications@github.com>
Reply-To: kubernetes/kubernetes <reply@reply.github.com>
To: kubernetes/kubernetes <kubernetes@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <kubernetes/kubernetes/pull/46456/c305555502@github.com>
In-Reply-To: <kubernetes/kubernetes/pull/46456@github.com>
References: <kubernetes/kubernetes/pull/46456@github.com>
Subject: Re: [kubernetes/kubernetes] Add local storage (scratch space)
 allocatable support (#46456)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_593047b5a921b_26c23f8b208edc2c804a4";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: k8s-ci-robot
X-GitHub-Recipient: falenn
X-GitHub-Reason: subscribed
List-ID: kubernetes/kubernetes <kubernetes.kubernetes.github.com>
List-Archive: https://github.com/kubernetes/kubernetes
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000ab60a7f1e480e831b7596b9ab36e3ed1a01ede9bcf2da92cf00000001154809b592a169ce0dcb7801@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAq2CrdIzuxk6g7nqElHkgwdOvlz1QYYks5r_u21gaJpZM4NmzRq>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: dev@imwiz.com
X-SG-EID: APO41b8ovafPb3SK9rw3vKQlNXZn2iegq+QyFRTyxCAh6308aRgfVnY7jwUo2S1XEaliMKYcwmkQYr
 1q5hEpPgniJ0PBEv2TpgmtA+cftD0gShWh5jjuAJrtPCZjLeUHvvRsreT2TmdCmXPLXfzq24w3N66c
 sHAEMsPnxvri8VI/EzJeqPUF8cfAzdTFl1sUE9nFea5sKwugK8ByS8A7GA==
X-Spam-Status: No, score=-1.7
X-Spam-Score: -16
X-Spam-Bar: -
X-Ham-Report: Spam detection software, running on the system "box969.bluehost.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  @jingxu97: The following test(s) **failed**: Test name | Commit
    | Details | Rerun command --- | --- | --- | --- pull-kubernetes-kubemark-e2e-gce
    | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-kubemark-e2e-gce/33040/)
    | `@k8s-bot pull-kubernetes-kubemark-e2e-gce test this` pull-kubernetes-federation-e2e-gce
    | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-federation-e2e-gce/7885/)
    | `@k8s-bot pull-kubernetes-federation-e2e-gce test this` pull-kubernetes-node-e2e
    | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-node-e2e/33174/)
    | `@k8s-bot pull-kubernetes-node-e2e test this` pull-kubernetes-verify |
   ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-verify/33046/)
    | `@k8s-bot pull-kubernetes-verify test this` pull-kubernetes-e2e-kops-aws
    | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-e2e-kops-aws/28278/)
    | `@k8s-bot pull-kubernetes-e2e-kops-aws test this` pull-kubernetes-e2e-gce-etcd3
    | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-e2e-gce-etcd3/33486/)
    | `@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this` pull-kubernetes-bazel
    | 552a4ac48ce46b307a3ae6a3bdf2d77f3184001c | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-bazel/24283/)
    | `@k8s-bot pull-kubernetes-bazel test this` pull-kubernetes-unit | 552a4ac48ce46b307a3ae6a3bdf2d77f3184001c
    | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-unit/34143/)
    | `@k8s-bot pull-kubernetes-unit test this` [...] 
 
 Content analysis details:   (-1.7 points, 4.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -0.0 RP_MATCHES_RCVD        Envelope sender domain matches handover relay domain
 -0.0 SPF_PASS               SPF: sender matches SPF record
 -0.0 RCVD_IN_MSPIKE_H3      RBL: Good reputation (+3)
                             [167.89.101.201 listed in wl.mailspike.net]
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.0 RCVD_IN_MSPIKE_WL      Mailspike good senders
 -1.6 AWL                    AWL: Adjusted score from AWL reputation of From: address
X-Spam-Flag: NO

----==_mimepart_593047b5a921b_26c23f8b208edc2c804a4
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@jingxu97: The following test(s) **failed**:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pull-kubernetes-kubemark-e2e-gce | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-kubemark-e2e-gce/33040/) | `@k8s-bot pull-kubernetes-kubemark-e2e-gce test this`
pull-kubernetes-federation-e2e-gce | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-federation-e2e-gce/7885/) | `@k8s-bot pull-kubernetes-federation-e2e-gce test this`
pull-kubernetes-node-e2e | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-node-e2e/33174/) | `@k8s-bot pull-kubernetes-node-e2e test this`
pull-kubernetes-verify | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-verify/33046/) | `@k8s-bot pull-kubernetes-verify test this`
pull-kubernetes-e2e-kops-aws | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-e2e-kops-aws/28278/) | `@k8s-bot pull-kubernetes-e2e-kops-aws test this`
pull-kubernetes-e2e-gce-etcd3 | ae28ad957c635cc599d395330681eb2fc998eb10 | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-e2e-gce-etcd3/33486/) | `@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this`
pull-kubernetes-bazel | 552a4ac48ce46b307a3ae6a3bdf2d77f3184001c | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-bazel/24283/) | `@k8s-bot pull-kubernetes-bazel test this`
pull-kubernetes-unit | 552a4ac48ce46b307a3ae6a3bdf2d77f3184001c | [link](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46456/pull-kubernetes-unit/34143/) | `@k8s-bot pull-kubernetes-unit test this`

[Full PR test history](https://k8s-gubernator.appspot.com/pr/46456). [Your PR dashboard](https://k8s-gubernator.appspot.com/pr/jingxu97). Please help us cut down on flakes by [linking to](https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-request-commands.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/kubernetes/kubernetes/pull/46456#issuecomment-305555502
----==_mimepart_593047b5a921b_26c23f8b208edc2c804a4
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: quoted-printable

<p><a href=3D"https://github.com/jingxu97" class=3D"user-mention">@jingxu97=
</a>: The following test(s) <strong>failed</strong>:</p>
<table>
<thead>
<tr>
<th>Test name</th>
<th>Commit</th>
<th>Details</th>
<th>Rerun command</th>
</tr>
</thead>
<tbody>
<tr>
<td>pull-kubernetes-kubemark-e2e-gce</td>
<td><a href=3D"https://github.com/kubernetes/kubernetes/commit/ae28ad957c63=
5cc599d395330681eb2fc998eb10" class=3D"commit-link"><tt>ae28ad9</tt></a></t=
d>
<td><a href=3D"https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/=
pr-logs/pull/46456/pull-kubernetes-kubemark-e2e-gce/33040/">link</a></td>
<td><code>@k8s-bot pull-kubernetes-kubemark-e2e-gce test this</code></td>
</tr>
<tr>
<td>pull-kubernetes-federation-e2e-gce</td>
<td><a href=3D"https://github.com/kubernetes/kubernetes/commit/ae28ad957c63=
5cc599d395330681eb2fc998eb10" class=3D"commit-link"><tt>ae28ad9</tt></a></t=
d>
<td><a href=3D"https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/=
pr-logs/pull/46456/pull-kubernetes-federation-e2e-gce/7885/">link</a></td>
<td><code>@k8s-bot pull-kubernetes-federation-e2e-gce test this</code></td>
</tr>
<tr>
<td>pull-kubernetes-node-e2e</td>
<td><a href=3D"https://github.com/kubernetes/kubernetes/commit/ae28ad957c63=
5cc599d395330681eb2fc998eb10" class=3D"commit-link"><tt>ae28ad9</tt></a></t=
d>
<td><a href=3D"https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/=
pr-logs/pull/46456/pull-kubernetes-node-e2e/33174/">link</a></td>
<td><code>@k8s-bot pull-kubernetes-node-e2e test this<

davidopp commented:
/approve

fejta commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref https://github.com/kubernetes/test-infra/pull/2931

jingxu97 commented:
@k8s-bot pull-kubernetes-unit test this

jingxu97 commented:
@k8s-bot pull-kubernetes-unit test this

jingxu97 commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this
